cosmwasm/contracts/ibc-translator/src/reply.rs

```diff
@@ -0,0 +1,174 @@
+        .contract
+        .context("no contract in response, we should never get here")?;
+
+    // load interim state
+    let transfer_info = CURRENT_TRANSFER
+        .load(deps.storage)
+        .context("failed to load current transfer from storage")?;
+
+    // delete interim state
+    CURRENT_TRANSFER.remove(deps.storage);
+
+    // deserialize payload into the type we expect
+    let payload: GatewayIbcTokenBridgePayload = serde_json_wasm::from_slice(&transfer_info.payload)
+        .context("failed to deserialize transfer payload")?;
+
+    match payload {
+        GatewayIbcTokenBridgePayload::Simple {
+            chain,
+            recipient,
+            fee: _,
+            nonce: _,
+        } => {
+            let recipient_decoded = String::from_utf8(recipient.to_vec())
+                .context(format!("failed to convert {recipient} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                recipient_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                None,
+            )
+        }
+        GatewayIbcTokenBridgePayload::ContractControlled {
+            chain,
+            contract,
+            payload,
+            nonce: _,
+        } => {
+            let contract_decoded = String::from_utf8(contract.to_vec())
+                .context(format!("failed to convert {contract} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                contract_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                Some(payload),
+            )
+        }
+    }
+}
+pub fn convert_cw20_to_bank_and_send(
+    deps: DepsMut,
+    env: Env,
+    recipient: String,
+    amount: u128,
+    cw20_contract_addr: String,
+    chain_id: u16,
+    payload: Option<Binary>,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    deps.api
+        .addr_validate(&cw20_contract_addr)
+        .context(format!("invalid contract address {cw20_contract_addr}"))?;
+
+    // convert contract address into base64
+    let subdenom = contract_addr_to_base58(deps.as_ref(), cw20_contract_addr.clone())?;
+    // format the token factory denom
+    let tokenfactory_denom = "factory/".to_string()
+        + env.contract.address.to_string().as_ref()
+        + "/"
+        + subdenom.as_ref();
+
+    let mut response: Response<TokenFactoryMsg> = Response::new();
+
+    // check contract storage see if we've created a denom for this cw20 token yet
+    // if we haven't created the denom, then create the denom
+    // info.sender contains the cw20 contract address
+    if !CW_DENOMS.has(deps.storage, cw20_contract_addr.clone()) {
+        // call into token factory to create the denom
+        let create_denom = SubMsg::new(TokenMsg::CreateDenom {
+            subdenom,
+            metadata: None,
```

Comment:
Is the metadata retrievable via some sort of explorer? Would it be worth it to put token info here?

Reply:
Currently, if a user wanted to get metadata on one of these tokens, they'd need to:
1. Convert the base58 subdenom into the bech32 cw20 contract address
2. Query the cw20 contract address to get token metadata

We could set token metadata here, I think it'd just require an extra cross-contract query to get the cw20 metadata from the cw20 contract. I don't know if any cosmos explorers automatically integrate denom metadata. 

@misko9 what are your thoughts?

Reply:
I haven't seen any explorers use tokenfactory's metadata yet either, but I think it's a good idea to add it. @nik-suri, can we sync on the correct mapping of CW20 metadata to tokenfactory metadata since it is not all 1:1?

Reply:
@misko9 and I discussed offline. It's useful to populate the tokenfactory metadata to prepare for a future where clients such as wallets and explorers can show the correct decimals for the ibc tokens. 

Since cosmos clients interpret decimals using a `base_denom` and `denom` structure (where `base_denom` is used as reference to compute decimals), we'll populate the tokenfactory metadata with 2 denom units - 
1. `base_denom` where the name is `base_<symbol>`
2. `denom` where the name is `<symbol>` and the exponent equals the number of decimals in for the cw20 token.